src/Workspaces/Core/Portable/Workspace/Solution/SolutionState_Checksum.cs

```diff
@@ -156,10 +157,22 @@ public async Task<Checksum> GetChecksumAsync(ProjectId projectId, CancellationTo
                     var analyzerReferenceChecksums = ChecksumCache.GetOrCreate<ChecksumCollection>(AnalyzerReferences,
                         _ => new ChecksumCollection(AnalyzerReferences.SelectAsArray(r => serializer.CreateChecksum(r, cancellationToken))));
 
-                    var projectChecksums = await Task.WhenAll(projectChecksumTasks).ConfigureAwait(false);
+                    var allResults = await Task.WhenAll(projectChecksumTasks).ConfigureAwait(false);
+                    using var _1 = ArrayBuilder<ProjectId>.GetInstance(projectChecksumTasks.Length, out var projectIds);
+                    using var _2 = ArrayBuilder<Checksum>.GetInstance(projectChecksumTasks.Length, out var projectChecksums);
+                    foreach (var projectStateChecksums in allResults)
+                    {
+                        if (projectStateChecksums != null)
+                        {
+                            projectIds.Add(projectStateChecksums.ProjectId);
+                            projectChecksums.Add(projectStateChecksums.Checksum);
+                        }
+                    }
+
                     return new SolutionStateChecksums(
                         attributesChecksum,
-                        new ChecksumCollection(projectChecksums.WhereNotNull().ToImmutableArray()),
+                        projectIds.ToImmutableAndClear(),
```

Comment:
passing along the project ids, not just their checksums.  that way when we jhave an instance of SolutionStateChecksums we can scope requests down to a particular id when trying to retrieve a particular project checksum.